javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java

```diff
@@ -48,111 +52,192 @@ public static ClientMethodTemplate getInstance() {
+
+            // Client methods only add local variable instantiations when the parameter isn't passed by the caller,
+            // isn't always null, is an optional parameter that was omitted or is a constant that is either required
+            // or AutoRest isn't generating with optional constant as enums.
             if (!parameter.getFromClient()
-                    && !alwaysNull
-                    && ((addOptional && clientMethod.getOnlyRequiredParameters() && !parameter.getIsRequired())
-                    || (addConstant && parameter.getIsConstant() && (!settings.isOptionalConstantAsEnum() || parameter.getIsRequired())))) {
+                && !alwaysNull
+                && ((addOptional && optionalOmitted) || (addConstant && includeConstant))) {
                 String defaultValue = parameterClientType.defaultValueExpression(parameter.getDefaultValue());
-                function.line("final %s %s = %s;", parameterClientType, parameter.getParameterReference(), defaultValue == null ? "null" : defaultValue);
+                function.line("final %s %s = %s;", parameterClientType, parameter.getParameterReference(),
+                    defaultValue == null ? "null" : defaultValue);
             }
         }
     }
 
-    protected static void applyParameterTransformations(JavaBlock function, ClientMethod clientMethod, JavaSettings settings) {
+    /**
+     * Applies parameter transformations to the client method parameters.
+     *
+     * @param function The client method code block.
+     * @param clientMethod The client method.
+     * @param settings AutoRest generation settings.
+     */
+    protected static void applyParameterTransformations(JavaBlock function, ClientMethod clientMethod,
+        JavaSettings settings) {
         for (MethodTransformationDetail transformation : clientMethod.getMethodTransformationDetails()) {
             if (transformation.getParameterMappings().isEmpty()) {
                 // the case that this flattened parameter is not original parameter from any other parameters
                 ClientMethodParameter outParameter = transformation.getOutParameter();
                 if (outParameter.getIsRequired() && outParameter.getClientType() instanceof ClassType) {
-                    function.line("%1$s %2$s = new %1$s();",
-                            outParameter.getClientType(),
-                            outParameter.getName());
+                    function.line("%1$s %2$s = new %1$s();", outParameter.getClientType(), outParameter.getName());
                 } else {
-                    function.line("%1$s %2$s = null;",
-                            outParameter.getClientType(),
-                            outParameter.getName());
+                    // TODO (alzimmer): This is making the assumption that the client type is nullable.
+                    // Is that always the case? What if the client type is a primitive?
```

Comment:
Since the `if` condition checks that the parameter is required, it cannot be a primitive type here in the `else` block. Since all optional params should be non-primitive types. Also, when the param is required, the client type will be an instance of ClassType.